jacodb-panda-dynamic/src/main/kotlin/org/jacodb/panda/dynamic/ets/dto/convert.kt

```diff
@@ -0,0 +1,712 @@
+import org.jacodb.panda.dynamic.ets.base.EtsStaticFieldRef
+import org.jacodb.panda.dynamic.ets.base.EtsStmt
+import org.jacodb.panda.dynamic.ets.base.EtsStringConstant
+import org.jacodb.panda.dynamic.ets.base.EtsStringType
+import org.jacodb.panda.dynamic.ets.base.EtsSwitchStmt
+import org.jacodb.panda.dynamic.ets.base.EtsThis
+import org.jacodb.panda.dynamic.ets.base.EtsThrowStmt
+import org.jacodb.panda.dynamic.ets.base.EtsTupleType
+import org.jacodb.panda.dynamic.ets.base.EtsType
+import org.jacodb.panda.dynamic.ets.base.EtsTypeOfExpr
+import org.jacodb.panda.dynamic.ets.base.EtsUnaryOperation
+import org.jacodb.panda.dynamic.ets.base.EtsUnclearRefType
+import org.jacodb.panda.dynamic.ets.base.EtsUndefinedConstant
+import org.jacodb.panda.dynamic.ets.base.EtsUndefinedType
+import org.jacodb.panda.dynamic.ets.base.EtsUnionType
+import org.jacodb.panda.dynamic.ets.base.EtsUnknownType
+import org.jacodb.panda.dynamic.ets.base.EtsValue
+import org.jacodb.panda.dynamic.ets.base.EtsVoidType
+import org.jacodb.panda.dynamic.ets.base.UnaryOp
+import org.jacodb.panda.dynamic.ets.graph.EtsCfg
+import org.jacodb.panda.dynamic.ets.model.EtsClass
+import org.jacodb.panda.dynamic.ets.model.EtsClassImpl
+import org.jacodb.panda.dynamic.ets.model.EtsClassSignature
+import org.jacodb.panda.dynamic.ets.model.EtsField
+import org.jacodb.panda.dynamic.ets.model.EtsFieldImpl
+import org.jacodb.panda.dynamic.ets.model.EtsFieldSignature
+import org.jacodb.panda.dynamic.ets.model.EtsFieldSubSignature
+import org.jacodb.panda.dynamic.ets.model.EtsFile
+import org.jacodb.panda.dynamic.ets.model.EtsMethod
+import org.jacodb.panda.dynamic.ets.model.EtsMethodImpl
+import org.jacodb.panda.dynamic.ets.model.EtsMethodParameter
+import org.jacodb.panda.dynamic.ets.model.EtsMethodSignature
+import org.jacodb.panda.dynamic.ets.model.EtsMethodSubSignature
+
+class EtsMethodBuilder(
+    signature: EtsMethodSignature,
+    // Default locals count is args + this
+    localsCount: Int = signature.parameters.size + 1,
+    modifiers: List<String> = emptyList(),
+) {
+    private val etsMethod = EtsMethodImpl(signature, localsCount, modifiers)
+    private var freeLocal: Int = 0
+    private val currentStmts: MutableList<EtsStmt> = mutableListOf()
+
+    private fun loc(): EtsInstLocation {
+        return EtsInstLocation(etsMethod, currentStmts.size)
+    }
+
+    private var built: Boolean = false
+
+    fun build(cfgDto: CfgDto): EtsMethod {
+        require(!built) { "Method has already been built" }
+        val cfg = cfg2cfg(cfgDto)
+        etsMethod.cfg = cfg
+        built = true
+        return etsMethod
+    }
+
+    fun convertToEtsStmt(stmt: StmtDto) {
+        val newStmt = when (stmt) {
+            is UnknownStmtDto -> object : EtsStmt {
+                override val location: EtsInstLocation = loc()
+
+                override fun toString(): String = "Unknown(${stmt.stmt})"
+
+                // TODO: equals/hashCode ???
```

Comment:
Implement it?

Reply:
I am not sure whether we even need it, or how to "properly" implement it (hence the *triple* question mark in the comment, that's rare). If we override equals (delegate to signature comparison), then different "unknown statements" will get lost after you put them in a set. That might be a good behavior for "normal" statements, but "unknown statements" are indicators of "something is incomplete/wrong", so we probably do not want to lose them in sets. The default ref-equals seems to be OK for them.